File: test_scripts/python/test_client.py
import sys
import getopt
import json


def getErrorMessage(error):
    error_type = type(error).__name__
    message = error.__str__()
    if error_type == 'ServerError':
        message = error.data
    return message


def main(argv):
    sys.path.append('./')
    tests_filepath = None
    endpoint = None
    token = None
    async_job_check_time_ms = None
    try:
        opts, args = getopt.getopt(
            argv,
            "ht:e:o:a:",
            ["help", "tests=", "endpoint=", "token=", "asyncchecktime="]
        )
    except getopt.GetoptError:
        print('Please use "test_client.py -h" or "test_client.py --help" for help')
        sys.exit(2)
    for opt, arg in opts:
        if opt in ("-h", "--help"):
            print((
                "test_client.py --tests "
                "<json_file> --endpoint <url> "
                "[--token <kbase_token> [--asyncchecktime <ms>]]"
            ))
            sys.exit()
        elif opt in ("-t", "--tests"):
            tests_filepath = arg
        elif opt in ("-e", "--endpoint"):
            endpoint = arg
        elif opt in ("-o", "--token"):
            token = arg
        elif opt in ("-a", "--asyncchecktime"):
            async_job_check_time_ms = int(arg)
    fh = open(tests_filepath)
    tests_json = json.load(fh)
    module_file = tests_json['package']
    class_name = tests_json['class']
    module = __import__(module_file)
    client_class = getattr(module, class_name)
    for test in tests_json['tests']:
        client_instance = None
        if 'auth' in test and test['auth']:
            if async_job_check_time_ms:
                client_instance = client_class(
                    url=endpoint,
                    token=token,
                    async_job_check_time_ms=async_job_check_time_ms
                )
            else:
                client_instance = client_class(url=endpoint, token=token)
        else:
            client_instance = client_class(url=endpoint, ignore_authrc=True)
        method_name = test['method']
        params = test['params']
        method_instance = getattr(client_instance, method_name)
        outcome = test['outcome']
        expected_status = outcome['status']
        ret = None
        error = None
        try:
            ret = method_instance(*params, context={})
        except Exception as ex:
            error = ex
        if expected_status == 'pass' or expected_status == 'nomatch':
            if error:
                print('Unexpected error for method=' + method_name
                      + ", params=" + json.dumps(params) + ":")
                print(getErrorMessage(error))
                sys.exit(1)
            if expected_status == 'pass':
                expected_ret = params
                if len(expected_ret) == 1:
                    expected_ret = expected_ret[0]
                elif len(expected_ret) == 0:
                    expected_ret = None
                expected_ret_json = json.dumps(expected_ret, sort_keys=True)
                actual_ret_json = json.dumps(ret, sort_keys=True)
                if expected_ret_json != actual_ret_json:
                    print('Output doesn\'t match input for method='
                          + method_name + ", params=" + json.dumps(params) + ":")
                    print('Expected output=' + expected_ret_json)
                    print('Observed output=' + actual_ret_json)
                    sys.exit(1)
        elif expected_status == 'fail':
            if error:
                if 'error' in outcome:
                    expected_error_fragments = outcome['error']
                    message = getErrorMessage(error)
                    wrong_fragment = None
                    for error_fragment in expected_error_fragments:
                        if error_fragment not in message:
                            wrong_fragment = error_fragment
                            break
                    if wrong_fragment:
                        print('Expected error fragment wasn\'t found in error happened for method='
                              + method_name + ", params=" + json.dumps(params) + ":")
                        print('Missing fragment: ' + wrong_fragment)
                        print('Full error: ' + message)
                        sys.exit(1)
            else:
                print('Error was expected but it hasn\'t happen for method='
                      + method_name + ", params=" + json.dumps(params)
                      + " -> ret=" + json.dumps(ret))
                sys.exit(1)
        else:
            print('Unsupported outcome status: ' + expected_status)
            sys.exit(1)


if __name__ == "__main__":
    main(sys.argv[1:])

File: test_scripts/perl/test-client.pl
#
# Automatic test rig for KBase perl client generation.
#
use strict;
use warnings;

use JSON;
use Data::Dumper;
use Test::More;
use Getopt::Long;

my $DESCRIPTION = "
USAGE
      perl test-client.pl [options]

DESCRIPTION
      Conducts an automatic test of a perl client against the specified endpoint using
      methods and parameters indicated in the tests config file.

      --tests [filename] config file for tests
      --endpoint [url]   endpoint of the server to test
      --token [token]    token for testing authenticated calls
      --asyncchecktime [ms] time client waits every cycle of checking state of async methods
      -h, --help         display this help message, ignore all arguments
";

# first parse options to get the testconfig file
my $tests_filename;
my $verbose;
my $endpoint;

my $token;
my $async_job_check_time_ms;

my $help;

my $opt = GetOptions(
    "tests=s"          => \$tests_filename,
    "verbose|v"        => \$verbose,
    "endpoint=s"       => \$endpoint,
    "token=s"          => \$token,
    "asyncchecktime=i" => \$async_job_check_time_ms,
    "help|h"           => \$help,
);

if ( $help ) {
    print $DESCRIPTION. "\n";
    exit 0;
}

# endpoint must be defined
if ( !$endpoint ) {
    fail( "endpoint parameter must be defined" );
    done_testing();
    exit 1;
}

# tests must be defined
if ( !$tests_filename ) {
    fail( "tests parameter must be defined" );
    done_testing();
    exit 1;
}

#parse the tests
open( my $fh, "<", $tests_filename );
my $tests_string = '';
while ( my $line = <$fh> ) {
    chomp $line;
    $tests_string .= $line;
}
close( $fh );
my $tests_json = JSON->new->decode( $tests_string );

my $tests         = $tests_json->{ tests };
my $client_module = $tests_json->{ package };

if ( !$tests ) {
    fail( "tests array not defined in test config file" );
    done_testing();
    exit 1;
}
if ( !$client_module ) {
    fail( "client module not defined in test config file" );
    done_testing();
    exit 1;
}

# make sure we can import the module
my $json = JSON->new->canonical;
use_ok( $client_module );

#instantiate an authenticated client and a nonauthenticated client
my $nonauthenticated_client = $client_module->new( $endpoint, ignore_kbase_config => 1 );
ok( defined( $nonauthenticated_client ), "instantiating nonauthenticated client" );

my $authenticated_client;
if ( $token ) {
    if ( $async_job_check_time_ms ) {
        $authenticated_client = $client_module->new(
            $endpoint,
            token                   => $token,
            async_job_check_time_ms => $async_job_check_time_ms
        );
    }
    else {
        $authenticated_client = $client_module->new( $endpoint, token => $token );
    }
    ok( defined( $authenticated_client ), "instantiating authenticated client" );
}

# loop over each test case and run it against the server.  We create a new client instance
# for each test
foreach my $test ( @{ $tests } ) {
    my $client;
    if ( $test->{ 'auth' } ) {
        if ( $authenticated_client ) {
            $client = $authenticated_client;
        }
        else {
            fail( "authenticated call declared, but no user and password set" );
            done_testing();
            exit 1;
        }

    }
    else {
        $client = $nonauthenticated_client;
    }
    ok( defined( $client ), "instantiating client" );

    my $method  = $test->{ method };
    my $params  = $test->{ params };
    my $outcome = $test->{ outcome };
    my $useScalarContext;
    if ( $test->{ context } && $test->{ context } eq 'scalar' ) {
        $useScalarContext = 1;
    }

    ok( $client->can( $method ),
        ' ----------------- method "' . $method . '" exists ------------------------' );
    if ( $client->can( $method ) ) {
        my ( @result, $scalarResult, $failed );
        {
            no strict "refs";
            eval {
                if ( $useScalarContext ) {
                    $scalarResult = $client->$method( @{ $params } );
                }
                else {
                    @result = $client->$method( @{ $params } );
                }
            };
            if ( $@ ) {
                $failed = 1;
                if ( $outcome->{ status } eq 'fail' ) {
                    pass( 'expected failure, and yes it failed' );
                }
                else {
                    fail( 'did not expect to fail, but it did' );
                    print STDERR
                        "Failing test of '$method', expected to pass but error thrown:\n";
                    print STDERR $@->{ message } . "\n";
                    if ( defined( $@->{ status_line } ) ) {
                        print STDERR $@->{ status_line } . "\n";
                    }
                }

                # check that the error message contains the right message
                if ( $outcome->{ error } ) {
                    my $returnedErrorMsg = $@->{ message };
                    foreach my $e ( @{ $outcome->{ error } } ) {
                        if ( index( $returnedErrorMsg, $e ) != -1 ) {
                            pass( "returned error message contains $e" );
                        }
                        else {
                            fail( "returned error message did not contain $e" );
                            print STDERR
                                "Failing test of '$method', expected error to contain message, but it didn't:\n";
                            print STDERR "  looking for: '$e'\n";
                            print STDERR "  got: '$returnedErrorMsg'\n";
                        }
                    }
                }

                # could do more checks here for different failure modes

            }
        }
        if ( $outcome->{ status } eq 'pass' ) {
            pass( 'expected to run successfully, and it did' );
            my ( $serialized_params, $serialized_result );
            if ( $useScalarContext ) {
                $serialized_params = $json->encode( [ @{ $params }[ 0 ] ] );
                $serialized_result = $json->encode( [ $scalarResult ] );
            }
            else {
                $serialized_params = $json->encode( $params );
                $serialized_result = $json->encode( \@result );
            }

            ok(
                $serialized_params eq $serialized_result,
                "response matches input parameters"
            );

            if ( $serialized_params ne $serialized_result ) {
                print STDERR
                    "Failing test of '$method', expected input to match output, but they don't match:\n";
                print STDERR "  in:  " . $serialized_params . "\n";
                print STDERR "  out: " . $serialized_result . "\n";
                print STDERR "\n";
            }
        }
        elsif ( $outcome->{ status } eq 'nomatch' ) {
            pass( 'expected to run successfully, and it did' );
            my ( $serialized_params, $serialized_result );
            if ( $useScalarContext ) {
                ok( $scalarResult, "recieved a response in scalar context" );
                $serialized_params = $json->encode( [ @{ $params }[ 0 ] ] );
                $serialized_result = $json->encode( [ $scalarResult ] );
            }
            else {
                ok( @result, "recieved a response in array context" );
                $serialized_params = $json->encode( $params );
                $serialized_result = $json->encode( \@result );
            }

            ok(
                $serialized_params ne $serialized_result,
                "response does NOT match input parameters"
            );

            if ( $serialized_params eq $serialized_result ) {
                print STDERR
                    "Failing test of '$method', expected input to NOT match output, but they did match:\n";
                print STDERR "  in/out:  " . $serialized_params . "\n";
                print STDERR "\n";
            }
        }
        elsif ( $outcome->{ status } eq 'fail' ) {
            if ( !$failed ) {
                fail( 'expected to fail, but it ran successfully' );
            }
        }
        else {
            fail(     'expected outcome set to "'
                    . $outcome->{ status }
                    . '", but that is not recognized.  Outcome can only be: pass | fail | nomatch'
            );
        }
    }
}

done_testing();


